File: tests/integration/test_research_host.cpp
#include "doctest.h"

#include "unit/Python/PythonTestSupport.h"

using namespace oryx;
using namespace oryx::test;

#ifdef OX_ENABLE_PYTHON

namespace
{

std::string expected_version()
{
    return std::to_string(VERSION_MAJOR) + "." + std::to_string(VERSION_MINOR) + "." + std::to_string(VERSION_PATCH);
}

} // namespace

TEST_SUITE("integration")
{

TEST_CASE("research host: import oryx works standalone")
{
    if (!can_run_python_extension())
    {
        return;
    }
    CHECK(run_python("import oryx; oryx.init()") == 0);
}

TEST_CASE("research host: the embedded host has __version__ but no init()")
{
    std::string output = run_oryx_script("mark(str(hasattr(oryx, 'init')) + ' ' + oryx.__version__)\n");

    CHECK(output == "False " + expected_version());
}

TEST_CASE("research host: oryx.__version__ matches the C++ version")
{
    if (!can_run_python_extension())
    {
        return;
    }
    CHECK(run_python("import oryx, sys; sys.exit(0 if oryx.__version__ == '" + expected_version() + "' else 1)") == 0);
}

TEST_CASE("research host: the context cannot be re-created once the atexit teardown ran")
{
    if (!can_run_python_extension())
    {
        return;
    }
    int status = run_python(
        "import atexit, os\n"
        "def late():\n"
        "    import oryx\n"
        "    try:\n"
        "        class Late(oryx.Strategy, id='late'):\n"
        "            def decide(self, context): return 0\n"
        "    except oryx.OryxError as e:\n"
        "        os._exit(0 if 'shutting down' in str(e) else 4)\n"
        "    os._exit(3)\n"
        "atexit.register(late)\n"
        "import oryx\n"
        "oryx.init()\n"
        "class Early(oryx.Strategy, id='early'):\n"
        "    def decide(self, context): return 0\n");
    CHECK(status == 0);
}

TEST_CASE("research host: benchmark(memory=True) counts Oryx's own allocations and balances them")
{
    if (!can_run_python_extension())
    {
        return;
    }
    std::filesystem::path scripts_dir = repo_file("Oasis/scripts");
    int status = run_python(
        "import sys; sys.path.insert(0, r'" + scripts_dir.string() + "')\n"
        "import oryx, nim\n"
        "oryx.init()\n"
        "assert oryx.benchmark.benchmark('nim', ['random', 'random'], games=5).memory is None\n"
        "m = oryx.benchmark.benchmark('nim', ['random', 'random'], games=20, memory=True).memory\n"
        "sys.exit(0 if m.allocation_count > 0 and m.allocation_count == m.deallocation_count and m.live_bytes == 0 else 1)");
    CHECK(status == 0);
}

TEST_CASE("research host: init() loads the scripts of the nearest oryx.yaml, an explicit settings file, or nothing")
{
    if (!can_run_python_extension())
    {
        return;
    }
    TempDir dir;
    dir.write("project/scripts/pile.py", "import oryx\nclass Pile(oryx.Game, id='pile'):\n    num_players = 2\n    def new_initial_state(self): return None\n");
    dir.write("project/oryx.yaml", "scripting:\n  roots:\n    - scripts\n");
    std::filesystem::create_directories(dir.path() / "project" / "notebooks");
    std::filesystem::create_directories(dir.path() / "elsewhere");

    std::string nested = (dir.path() / "project" / "notebooks").string();
    std::string elsewhere = (dir.path() / "elsewhere").string();
    std::string settings = (dir.path() / "project" / "oryx.yaml").string();

    CHECK(run_python("import os, sys; os.chdir(r'" + nested + "'); import oryx; oryx.init(); sys.exit(0 if 'pile' in oryx.list_games() else 1)") == 0);
    CHECK(run_python("import os, sys; os.chdir(r'" + elsewhere + "'); import oryx; oryx.init(); sys.exit(0 if oryx.list_games() == [] else 1)") == 0);
    CHECK(run_python("import os, sys; os.chdir(r'" + elsewhere + "'); import oryx; oryx.init(r'" + settings + "'); sys.exit(0 if 'pile' in oryx.list_games() else 1)") == 0);
    CHECK(run_python("import sys, oryx\ntry:\n    oryx.init('/no/such/oryx.yaml')\nexcept oryx.SettingsError:\n    sys.exit(0)\nsys.exit(1)") == 0);
}

TEST_CASE("research host: a scripted game and simulate() exit cleanly under the debug allocator")
{
    if (!can_run_python_extension())
    {
        return;
    }
    std::filesystem::path scripts_dir = repo_file("Oasis/scripts");
    int status = run_python(
        "import sys; sys.path.insert(0, r'" + scripts_dir.string() + "')\n"
        "import oryx, nim\n"
        "oryx.init()\n"
        "result = oryx.simulate('nim', ['random', 'random'], games=20)\n"
        "sys.exit(0 if result.matches == 20 else 1)",
        "PYTHONMALLOC=debug ");
    CHECK(status == 0);
}

TEST_CASE("research host: assertion hook raises instead of trapping")
{
    if (!can_run_python_extension())
    {
        return;
    }
    int status = run_python(
        "import oryx, sys; oryx.init();\n"
        "try:\n"
        "    oryx.debug.check(False, 'x')\n"
        "except oryx.errors.OryxAssertionError:\n"
        "    sys.exit(0)\n"
        "sys.exit(1)");
    CHECK(status == 0);
}

TEST_CASE("research host: interruptible batch raises when the interpreter is interrupted mid-batch")
{
    if (!can_run_python_extension())
    {
        return;
    }
    // Bare `import oryx` registers no C++ game, so this interrupts a scripted one: the interrupt must pass through its callback untouched.
    std::filesystem::path scripts_dir = repo_file("Oasis/scripts");
    std::string code =
        "import sys; sys.path.insert(0, r'" + scripts_dir.string() + "')\n"
        "import oryx, nim, signal, threading, _thread\n"
        "oryx.init()\n"
        "signal.signal(signal.SIGINT, signal.default_int_handler)\n"
        "threading.Timer(0.05, _thread.interrupt_main).start()\n"
        "try:\n"
        "    oryx.simulate('nim', ['random', 'random'], games=2_000_000)\n"
        "except KeyboardInterrupt:\n"
        "    sys.exit(0)\n"
        "sys.exit(1)";
    int status = run_python(code);
    CHECK(status == 0);
}

} // TEST_SUITE("integration")

#endif // OX_ENABLE_PYTHON

File: tests/python/test_research_host.py
import os
import subprocess
import sys
from pathlib import Path

import oryx
import pytest

_SCRIPTS_DIR = Path(__file__).resolve().parents[2] / "Oasis" / "scripts"


def _run(code, **kwargs):
    return subprocess.run([sys.executable, "-c", code], **kwargs)


def test_import_oryx_works_standalone():
    assert _run("import oryx; oryx.init()").returncode == 0


def test_the_context_cannot_be_recreated_once_the_atexit_teardown_ran():
    code = """
import atexit, os


def late():
    import oryx
    try:
        class Late(oryx.Strategy, id="late"):
            def decide(self, context):
                return 0
    except oryx.OryxError as e:
        os._exit(0 if "shutting down" in str(e) else 4)
    os._exit(3)


atexit.register(late)
import oryx

oryx.init()


class Early(oryx.Strategy, id="early"):
    def decide(self, context):
        return 0
"""
    assert _run(code).returncode == 0


def test_benchmark_memory_true_counts_oryxs_own_allocations_and_balances_them():
    assert oryx.benchmark.benchmark("nim", ["random", "random"], games=5).memory is None
    m = oryx.benchmark.benchmark("nim", ["random", "random"], games=20, memory=True).memory
    assert m.allocation_count > 0
    assert m.allocation_count == m.deallocation_count
    assert m.live_bytes == 0


def test_init_loads_the_scripts_of_the_nearest_oryx_yaml_an_explicit_settings_file_or_nothing(tmp_path):
    (tmp_path / "project" / "scripts").mkdir(parents=True)
    (tmp_path / "project" / "scripts" / "pile.py").write_text(
        "import oryx\n\n\nclass Pile(oryx.Game, id='pile'):\n"
        "    num_players = 2\n\n    def new_initial_state(self):\n        return None\n"
    )
    (tmp_path / "project" / "oryx.yaml").write_text("scripting:\n  roots:\n    - scripts\n")
    nested = tmp_path / "project" / "notebooks"
    nested.mkdir()
    elsewhere = tmp_path / "elsewhere"
    elsewhere.mkdir()
    settings = tmp_path / "project" / "oryx.yaml"

    assert (
        _run("import sys, oryx; oryx.init(); sys.exit(0 if 'pile' in oryx.list_games() else 1)", cwd=nested).returncode
        == 0
    )
    assert (
        _run("import sys, oryx; oryx.init(); sys.exit(0 if oryx.list_games() == [] else 1)", cwd=elsewhere).returncode
        == 0
    )
    assert (
        _run(
            f"import sys, oryx; oryx.init({str(settings)!r}); sys.exit(0 if 'pile' in oryx.list_games() else 1)",
            cwd=elsewhere,
        ).returncode
        == 0
    )
    assert (
        _run(
            "import sys, oryx\ntry:\n    oryx.init('/no/such/oryx.yaml')\nexcept oryx.SettingsError:\n    sys.exit(0)\nsys.exit(1)",
            cwd=elsewhere,
        ).returncode
        == 0
    )


def test_a_scripted_game_and_simulate_exit_cleanly_under_the_debug_allocator():
    code = (
        f"import sys; sys.path.insert(0, {str(_SCRIPTS_DIR)!r})\n"
        "import oryx, nim\n"
        "oryx.init()\n"
        "result = oryx.simulate('nim', ['random', 'random'], games=20)\n"
        "sys.exit(0 if result.matches == 20 else 1)\n"
    )
    assert _run(code, env={**os.environ, "PYTHONMALLOC": "debug"}).returncode == 0


def test_assertion_hook_raises_instead_of_trapping():
    with pytest.raises(oryx.errors.OryxAssertionError):
        oryx.debug.check(False, "x")


def test_interruptible_batch_raises_when_the_interpreter_is_interrupted_mid_batch():
    # Bare `import oryx` registers no C++ game, so this interrupts a scripted one: the
    # interrupt must pass through its callback untouched.
    code = (
        f"import sys; sys.path.insert(0, {str(_SCRIPTS_DIR)!r})\n"
        "import oryx, nim, signal, threading, _thread\n"
        "oryx.init()\n"
        "signal.signal(signal.SIGINT, signal.default_int_handler)\n"
        "threading.Timer(0.05, _thread.interrupt_main).start()\n"
        "try:\n"
        "    oryx.simulate('nim', ['random', 'random'], games=2_000_000)\n"
        "except KeyboardInterrupt:\n"
        "    sys.exit(0)\n"
        "sys.exit(1)\n"
    )
    assert _run(code).returncode == 0
